torch/testing/_internal/schema_check_tensor.py

```diff
@@ -43,14 +55,38 @@ def unwrap(e):
         def wrap(e):
             return cls(e) if isinstance(e, torch.Tensor) else e
 
-        global schema_check_recorded_ops
-        schema_check_recorded_ops.append(func.__name__)
+        def has_mutated(before, after):
+            return not torch.equal(before, after) if isinstance(before, torch.Tensor) and isinstance(after, torch.Tensor) else False
+
+        SchemaCheckTensor.recorded_ops.append(func.__name__)
+        schema, arguments = find_matching_schema(func, tree_map(unwrap, args), tree_map(unwrap, kwargs))
+        cloned_arguments = dict(zip(arguments.keys(), clone_inputs(arguments.values())))
         out = func(*tree_map(unwrap, args), **tree_map(unwrap, kwargs))
+
+        for argument in schema.arguments:
+            name = argument.name if argument.name != "self" else "input"
+            if arguments.get(name) is not None:
+                before = tree_flatten(arguments.get(name))[0]
+                after = tree_flatten(cloned_arguments.get(name))[0]
+                if (any([has_mutated(i, j) for i, j in zip(before, after)]) and not argument.is_mutable):
+                    raise RuntimeError(f"Argument {name} is not defined as mutable but was mutated")
+
         return tree_map(wrap, out)
 
-def reset_cache():
-    global schema_check_recorded_ops
-    schema_check_recorded_ops.clear()
+def find_matching_schema(op: Callable, args : Tuple[Any], kwargs : Dict[str, Any]):
```

Comment:
Maybe @eellison would know better, but I think you can actually skip all the indirection here around `get_signature_for_torch_op`.

The idea with `get_signature_for_torch_op` is that if you have a torch operator like `torch.add`, it actually represents a bundle of operator overloads that each have separate schemas (for add, it would be `add.Tensor`, `add.Scalar` and `add.out`).

If you only had the bundle, then you'd be forced to do these lookups. But you're inside of a `__torch_dispatch__` call, so you're actually in a better situation: the `func` object passed into `__torch_dispatch__` represents a single overload. For example, if you called `torch.add`, then the `func` by the time you hit `__torch_dispatch__` would be `torch.ops.aten.add.Tensor`.

Since those overload objects represent individual overloads, then they have a unique schema. You can access it directly with `torch.ops.aten.add.Tensor._schema`, which returns a `FunctionSchema` object. I think that all the information that you need should be on there, but if not, we should probably just expose it to python.

Reply:
@bdhirsh I think `._schema` is missing the information about which arguments map to which values specifically due to the possibility of having both kwargs and args out of order. Is there a way to find this mapping from the `FunctionSchema` object without the signature? Because part of how I was using `get_signature_for_torch_op` was to get this mapping using `bind`